### DISCUSSION TITLE
Discussion: Proposing TagProperty backwards-incompatible change

### DISCUSSION THREAD

**Discussion by Griatch** [evennia/evennia#3150] [Evennia Engine Development]
# TLDR 

I propose changing Evennia's `TagProperty` (introduced in Evennia 1.0) from being written as 

    tagkey = TagProperty("category") 

to instead be written as 

    category = TagProperty("tagkey")

This would be a backwards-incompatible change, but is anyone actually using `TagProperty` extensively at this point? 

(for more info: [docs on Tags](https://www.evennia.com/docs/latest/Components/Tags.html))

# Problem

Currently `evennia.typeclasses.tags.TagProperty` is mimicking how `AttributeProperty` works. For example: 

```python
from evennia import DefaultCharacter, TagProperty, AttributeProperty

class MyCharacterClass(DefaultCharacter): 

    hp = AttributeProperty(default=10) 

    warriors = TagProperty(category="guild")
    elf = TagProperty(category="species")

```
This makes sense for `AttributeProperty` - you get a new Attribute `hp` that starts out at 10. At first glance, the `TagProperty` seems to work fine too - this character will end up with a Tag `warriors` with category `guild` - that is, they are tagged to be a member of the warrior's guild. They are also tagged with a `species` of  `elf`.

The problem is that while you can easily change the Attribute by just doing `char.hp = 4` later, this doesn't really make sense for the tag equivalent. The `warriors` tag key is now the property on the class. The property itself _is_ the value of the tag. In practice this makes the class inflexible - every instance of this class will be elf warriors and you must make a subclass to make anything different. 

You _could_ do `char.elf = another_class`, to change the category of the tag, but that doesn't really make much sense in practice and is not how tags are meant to work. 

I feel this makes `TagProperty` pretty useless in practice, and a poorly thought-through feature on my part.

# Suggestion 

I think the problems with `TagProperty` goes away if one simply swaps the tagkey with category like this: 

```python
class MyCharacterClass(DefaultCharacter): 

    guild = TagProperty("warriors")
    species = TagProperty("elf")

```
The idea here is that the tag `category` describes what this tag is all about, whereas the tag `key` is the 'value' that we want to be more flexible with and want to change. It makes a lot more sense for us to do `char.guild = 'wizards'` when the user changes to a different guild later, or `char.species = 'demon'` when they make an evil pact.

## Drawbacks

Some drawbacks with the proposed solution

- With the current `TagProperty` (and with `char.tags.add()`) you can create tags with a category of `None`. This will not be possible with this change; you can't syntactically do `None = TagProperty("foo")` in Python and I don't think it's a good idea to introduce some arbitrary category name to represent `None`. So all tags created using `TagProperty` would have to have a category. I don't think this is a major problem, there are many other ways to create tags if you want no category. 
- This change may trick users into thinking tags work more like attributes than they do. "Re-assigning" a tag with a different tagkey may look like assigning a new value to an Attribute, but it will in fact create a new tag object (or reuse an existing one with the same key+category combination). The tag is still shared between all objects like all tags - this is just a different way of assigning it. 
- The main drawback is that this is a backwards incompatible change. It would break existing code. 

I don't know if anyone is using `TagProperty` extensively and really like the old way, hence why I invite discussion here.

**Comment by InspectorCaracal** [comment]
I don't use `TagProperty` myself because I find it really hard to apply to any useful situation as it's written, so my main focus reading through this proposal is to see whether the change would make it potentially useful for me or not. So, that said:

I've read this a couple times now and thought hard about it and I think this change *would* make it potentially more useful. I could see a useful application for it for my own games with zones - if you have a `zone` `TagProperty` in your base `Room` class, it'll map to whatever `zone` tag you want on it, and you can then easily "move" a room from one zone to the other by changing the `zone` `TagProperty`, as well as easily check the zone of a room.

One big potential issue that immediately jumped to mind, however, is what happens if (to use one of your examples) the game allows you to be in more than one guild, and there is a `guild` `TagProperty` on characters. What would `chara.guild` return? If it's a list of all tags in `guild`, what happens if you then assign `chara.guild` to a new string? Or to a list of strings?

> This change may trick users into thinking tags work more like attributes than they do.

This is a fair concern, but honestly, the `TagProperty` as it currently stands is already pretty confusing in how it actually works and how to use it and how it's similar-but-actually-totally-different from the `AttributeProperty` both in function and in usage. I think this change would not be any more confusing in that respect, possibly even less so.

**Reply by InspectorCaracal** [reply]
As a followup/addendum, I think that tag properties being attached to categories and providing a convenient property-style point of access for checking and changing the tags in that category would be really nice and really useful if it properly supports having zero or multiple tags within the category. e.g. to use the `guild` example again, if it lets you see that a character isn't in a guild at all (`chara.guild` returns/is set to `None`) or in a specific guild (`"warrior" in chara.guild`), and that would fit really smoothly into the property-as-category and with builders/code assigning or removing tags to an object in the normal ways.

But, I don't have any really solid thoughts about how the specifics of *assigning* the tag property could, or should, interface with that. I feel like the cleanest and most straightforward approach would be to only do simple assignments and reassignments - that is, you can't use a tag property to add or remove individual tags to a category. But I think it probably _should_ be able to handle being given either a single string (to set it to just one tag in the category) or to an iterable like a list or tuple (to set it to multiple tags within the category).

**Reply by Griatch** [reply]
Hm, good point about multiple tags under one category. In this sense, the current implementation of `TagProperty` does represent a more 1:1 tag representation - it's really one tag per property, which may be a little easier to reason around. It also severely limits the usability though.

**Today:** 

    warriors = TagProperty("guild")
    wizards = TagProperty("guild") 
    thieves = TagProperty("guild") 

These are all instances of `obj.tags.add("warriors", "guild")` etc under the thood. As mentioned, while this makes the database mapping clear, it's not very useful. Tags are mainly used to find memberships in data sets. I guess you could ask 'is this object a member of the warrior's guild?' and check `getattr(obj, "warriors")` ... but there is no way to easily ask the question 'which guild does this object belong to?' without knowing and looping over all available guilds - you'd need to go for a db query.

**Proposal**

    guild = TagProperty("warriors", "wizards", "thieves")

This would be equivalent to `obj.tags.batch_add(("warriors", "guild"), ("wizards", "guild"), ("thieves", "guild"))`. You could now easily answer which guilds the object is part of: 

    memberships = obj.guild

(plural form may be confusing, but it depends on what the category is called). This is equivalent to 
`memberships = obj.tags.get(category="guild")`, which would then return all tags/guilds the object is a member of. If you separately used `obj.tags.add("townsguard", "guild")`, the property would include that too (this is no different from how an `AttributeProperty` could be modified by `.db` or `.attributes.add` as well). 

**Updating an existing property**

    obj.guild = "assassins"

I think it makes the most syntactic sense if this removes all `guild`-categorized tags on `obj`, replacing it with the new assassins one. You'd do something like this for multiplets:

    obj.guild = ("assassins", "thieves", "wizards", "warriors")

This would add/update all given tags. Since a tag-key can only a string, there would be no problem to identify an iterable to add.  

    del obj.guild

I think this must delete all tags of that category. There'd be no way of deleting only one with this syntax. One could picture having `TagProperty.remove/add` operations, but then you are making a new `TagHandler` ... I'd say that if you want more fine-grained control you should use the existing `obj.tags.add/remove` to manipulate individual tags. 

----

Overall, I feel that the `TagProperty` is mostly for convenience when setting up classes; I think it'd be very nice to be able to make a class and just say that all members of this class should start being tagged as something in particular without having to do so in `at_object_creation`.

**Comment by TonyLB** [comment]
Agreed with Caracal that this doesn't (immediately) seem like it's marking out a distinct conceptual space.

Creating a TagProperty seems like it is meaningfully part of creating a _de facto_ enumeration of allowed values:  It makes sense to imagine a file of legal TagProperty values for (for instance) guild, and only ever assigning **those properties** to the "guild" key. By contrast, allowing somebody to type `species = TagProperty("warriors")` feels like a conceptual error that should raise warnings. So, it feels like (much as Python loves its duck-typing), some type-checking would be a value-add.

If the TagProperty then has some sense of "Here's where this can be applied," then there's a second (distinct) question of "Can this field contain one unique property, or a set of properties?" It's possible that rather than something as simple as `guild = TagProperty("warriors")`, it would be useful to be able to say `guild = SetOfTags("guild")` followed by `guild.addTag(TagProperty("warriors"))` (or, alternately, `species = UniqueTag("species")` followed by `species.setTag(TagProperty("elf"))`)

**Reply by InspectorCaracal** [reply]
I appreciate the vote of confidence but this isn't what I was saying at all. 😅

I think that handling the fundamentally allowable presence of multiple tags within one category is an important consideration that _could_ be a significant obstacle in implementing this change, but I otherwise think the proposal is better than the existing `TagProperty`.

I also think that defining what counts as a "valid" tag for a given category is wildly outside the scope of the concept.

**Reply by TonyLB** [reply]
Fair, and that makes perfect sense.

**Comment by Griatch** [comment]
@TonyLB 

Hm, if I understand you correctly,

    species = TagProperty("warriors")

is indeed 'conceptually wrong' in the sense that that 'warriors' is not a 'species', but controlling this (enum-like behavior) goes beyond the scope of this proposal. As a developer you can add any tag+category key to your objects and `TagProperty` is just a different way to do just that. If you wanted a context-aware validation like this for your game, you would need to add it yourself.

**Comment by Griatch** [comment]
I have now implemented this, but I didn't change the existing `TagProperty`, instead I added a new `TagCategoryProperty` that works as outlined in this thread (now merged to `main`): 

```python 
from evennia import TagCategoryProperty

class Sword(Object):

   damage_type = TagCategoryProperty("slashing", "piercing") 
   crafting_component = TagCategoryProperty("hilt", "pommel", "blade") 
```
